lib/gc/Transforms/Microkernel/ConvertLinalgToMicrokernel.cpp

```diff
@@ -0,0 +1,370 @@
+
+  BrgemmDims() = default;
+  BrgemmDims(int64_t bdA, int64_t ldA, int64_t mdA, int64_t bdB, int64_t ldB,
+             int64_t mdB)
+      : batchDimA(bdA), leadingDimA(ldA), minorDimA(mdA), batchDimB(bdB),
+        leadingDimB(ldB), minorDimB(mdB) {}
+};
+
+FailureOr<linalg::ContractionDimensions>
+customInferContractionDims(linalg::LinalgOp linalgOp) {
+  auto dims = linalg::inferContractionDims(linalgOp);
+  if (failed(dims))
+    return failure();
+  if (llvm::isa<linalgx::BatchReduceMatmulVnniOp>(linalgOp)) {
+    // For VnniOp, the K reduction dims (dim index 3 & 4) cannot be infered by
+    // linalg utils because they form complex affine in operand A; Manually add
+    // them here
+    dims->k.push_back(3);
+    dims->k.push_back(4);
+  }
+  return dims;
+}
+
+static bool isMatchingAffineResult(linalg::LinalgOp linalgOp, AffineExpr expr,
+                                   ArrayRef<unsigned> dimPos) {
+  // Expecting dimPos.size() == 1 for normal dim and == 2 for vnni dim
+  if (dimPos.size() > 2)
+    return false;
+
+  auto firstDim = getAffineDimExpr(dimPos[0], linalgOp.getContext());
+  if (dimPos.size() == 1)
+    return firstDim == expr;
+
+  // If not regular dim affine, check for VNNI format K affine
+  auto secondKPosDim = getAffineDimExpr(dimPos[1], linalgOp.getContext());
+  // An K affine result for VNNI should be this format:
+  // d{kPos[0]} * s{kPos[1]} + d{kPos[1]} (k0 * K_vnni + k1)
+  auto add = dyn_cast<AffineBinaryOpExpr>(expr);
+  if (!add)
+    return false;
+  if (add.getKind() != AffineExprKind::Add)
+    return false;
+  auto lhs = add.getLHS();
+  auto rhs = add.getRHS();
+  if (rhs != secondKPosDim)
+    return false;
+  auto mul = dyn_cast<AffineBinaryOpExpr>(lhs);
+  if (!mul || mul.getKind() != AffineExprKind::Mul || mul.getLHS() != firstDim)
+    return false;
+
+  auto cst_affine = dyn_cast<AffineConstantExpr>(mul.getRHS());
+  return cst_affine &&
+         (cst_affine.getValue() == 2 || cst_affine.getValue() == 4);
+}
+
+// Return the position of linalg loop `dim` in the domain of `operand`.
+static std::optional<unsigned> getPosInDomain(ArrayRef<unsigned> dimPos,
+                                              OpOperand *operand,
+                                              linalg::LinalgOp linalgOp) {
+  assert(operand->getOwner() == linalgOp);
+  auto map = linalgOp.getMatchingIndexingMap(operand);
+  for (unsigned i = 0, numResults = map.getNumResults(); i < numResults; i++) {
+    if (isMatchingAffineResult(linalgOp, map.getResult(i), dimPos))
+      return i;
+  }
+  return std::nullopt;
+}
+
+static FailureOr<BrgemmDims> inferBrgemmDims(linalg::LinalgOp linalgOp) {
+  using namespace mlir::structured_match;
+  auto validBrgemmMatcher = StructuredOpMatcher::make<linalg::LinalgOp>()
+                                .output(MatchAll(), HasStaticShape())
+                                .input(MatchAll(), HasStaticShape())
+                                .output(MatchAll(), HasStaticStrides())
```

Comment:
The stride check should be useless here as it only check `MemRefType` while the current lianlg op is still in tensor type

Reply:
useless stride check removed.